docs/proposals/004-com-support.md

```diff
@@ -0,0 +1,240 @@
+Our intended path includes things that are compatible with the COM philosophy, like reference-counted lifetime management and idiomatic use of result/error codes, but those features don't currently align with the more explicit style used by COM in C/C++.
+
+Related Work
+------------
+
+The .NET platform includes some support for allowing .NET `interface`s and COM interfaces to interoperate.
+TODO: Need to study this and learn how it works.
+
+Proposed Approach
+-----------------
+
+We propose to allow COM interfaces to be declared using the Slang `interface` construct, with an appropriate attribute or modifier:
+
+```
+[COM] interface IDevice
+{
+    ITexture createTexture(__read TextureDesc desc) throws HRESULT;
+
+    void setTexture(int index, ITexture texture);
+}
+```
+
+A declaration like the above will translate into output C++ along the lines of:
+
+```
+struct IDevice : public IUnknown
+{
+    virtual HRESULT SLANG_MCALL createTexture(TextureDesc const& desc, ITexture** _result) = 0;
+
+    virtual void SLANG_MCALL setTexture(int index, ITexture* texture) = 0;
+};
+```
+
+Key things to note:
+
+* The `[COM] interface` becomes a C++ `struct` that inherits from `IUnknown`
+* Methods defined in the `[COM] interface` become pure-virtual `SLANG_MCALL` methods in C++
+* Parametes/values of a `[COM] interface` type `IFoo` in Slang translate to `IFoo*` in C++
+* Methods that have a `throws HRESULT` clause are transformed to have an `HRESULT` return type and an output parameter for their result
+
+A Slang `class` can declare that it implements zero or more `[COM] interface`s. Code like this:
+
+```
+class MyTexture : ITexture
+{
+    // ...
+}
+
+class MyDevice : IDevice
+{
+    ITexture createTexture(__read TextureDesc desc) throws HRESULT
+    {
+        return ...;
+    }
+
+    void setTexture(int index, ITexture texture)
+    {
+        ...;
+    }
+}
+```
+
+translates into output C++ like this:
+
+```
+class MyTexture : public slang::Object, public ITexture
+{
+    // ...
+};
+
+class MyDevice : public slang::Object, public  IDevice
+{
+    virtual HRESULT QueryInterface(REFIID riid, void **ppvObject) { ... }
+    virtual ULONG AddRef() { ... }
+    virtual ULONG Release() { ... }
+
+    HRESULT createTexture(TextureDesc const& desc, ITexture** _result) SLANG_OVERRIDE
+    {
+        _result = ...;
+        return S_OK;
+    }
+
+    void setTexture(int index, ITexture* texture) SLANG_OVERRIDE
+    {
+        ...
+    }
+}
+```
+
+All Slang `class`es translate to C++ classes that inherit from `slang::Object` (equivalent to the `RefObject` type within the current Slang implementation).
+A `class` that inherits any `[COM]` interfaces includes an implementation of `IUnknown` plus the methods to override all the interface requirements.
+
+In ordinary code that makes use of `[COM] interface` types:
+
+```
+struct Stuff
+{
+    ITexture t;
```

Comment:
Hmmm. I guess we are saying [COM] types are *reference* types implicitly?
In Stuff does it own the pointer or not?

Reply:
> I guess we are saying [COM] types are reference types implicitly?

It is more accurate to say that only reference types can conform to a `[COM]` interface.

> In `Stuff` does it own the pointer or not?

In terms of memory management, `Stuff` retains a reference to whatever `t` refers to. Not that the answer there doesn't care whether `ITexture` is a `[COM]` interface or not. An `IWhatever` existential that holds a value of `class` type will always retain a reference.

> Presumably the [COM] interfaces don't/can't work on GPUs?

I think that is a fair assumption *for now*. I wouldn't rule anything out in the long run.

> The default for interface seems (?) to be it's a value type.

Correct, insofar as that is more or less the only behavior. An existential in Slang always behaves like the value of concrete type that it holds. If the underlying concrete type is a value type, the existential behaves like a value. If the underlying concrete type is a reference, the existential behaves like a reference.

Really, the simplest way to think of it is that *everything* is a value type, and "reference types" are just those types where the *value* is a reference/pointer to some storage.

> The class type seems like that would slang reference type. Having a modifier change the types behavior so dramatically seems 'surprising'.

It is important to note that the `[COM]` attribute does **not** cause the change in behavior. The difference you are observing is around whether the concrete type is a `struct` type or a `class` type (a distinction that people have accepted easily enough in C#).

> Something that made more clear the reference nature might be beneficial.

I don't think that can be made to work, because the same distinction would hold for *any* interface when using `class` types. In fact, the same basic thing is *also* true with generics:

```
struct Stuff<T>
{
    T t;
}
```

The programmer cannot tell at the time they declare `Stuff` whether `T` will be a value type (`struct`) or reference type (`class`). The important thing is that in 99% of cases it turns out they don't need to care.